gensim/similarity_learning/preprocessing/sl_vocab.py

```diff
@@ -246,12 +335,69 @@ def get_batch(self, batch_size=32):
-            X2[i * 2 + 1, :neg_doc_len], X2_len[i * 2 + 1] = neg_doc[:neg_doc_len], neg_doc_len
+            X2[i * 2, :pos_doc_len], X2_len[i *
+                                            2] = pos_doc[:pos_doc_len], pos_doc_len
+            X1[i * 2 + 1, :query_len], X1_len[i *
+                                              2 + 1] = query[:query_len], query_len
+            X2[i * 2 + 1, :neg_doc_len], X2_len[i * 2 +
+                                                1] = neg_doc[:neg_doc_len], neg_doc_len
 
         return X1, X1_len, X2, X2_len, Y
 
+    def jbatch_gen(self, batch_size=32):
+        """Generator to provide a batch of training sample of size batch_size
+        The batch provided is actually twice the size and will have alternate positive
+        and negative examples. CHECK does alternation even matter?
+
+        So, if batch_size is 16
+        the returned batch will be of size 32 like:
+        [positive_example, negative_example, positive_example, ...]
+
+        TOCHECK : maybe this behaviour shouldn't be like this
+
+        Parameters:
+        ----------
+        batch_size: int
+            provides of batches of that size
+        """
+
+        while True:
+            # Initialize the query, doc and relevance arrays to zero
+            X1 = np.zeros((batch_size * 2, self.text_maxlen), dtype=np.int32)
+            X1_len = np.zeros((batch_size * 2,), dtype=np.int32)
+
+            X2 = np.zeros((batch_size * 2, self.text_maxlen), dtype=np.int32)
+            X2_len = np.zeros((batch_size * 2,), dtype=np.int32)
+
+            Y = np.zeros((batch_size * 2,), dtype=np.int32)
+
+            # set alternate relevances to 1, starting with 1
+            Y[::2] = 1
+
+            # set all the values to pad words
+            X1[:] = self.pad_word_index
+            X2[:] = self.pad_word_index
+
+            num_samples = len(self.pair_list)
+
+            for offset in range(0, num_samples, batch_size):
+                batch_sample = self.pair_list[offset: offset + batch_size]
+                for i, (query, pos_doc, neg_doc) in enumerate(batch_sample):
+                    query_len = min(self.text_maxlen, len(query))
+                    pos_doc_len = min(self.text_maxlen, len(pos_doc))
+                    neg_doc_len = min(self.text_maxlen, len(neg_doc))
+
+                    X1[i * 2, :query_len], X1_len[i *
+                                                  2] = query[:query_len], query_len
+                    X2[i * 2, :pos_doc_len], X2_len[i *
+                                                    2] = pos_doc[:pos_doc_len], pos_doc_len
+                    X1[i * 2 + 1, :query_len], X1_len[i *
+                                                      2 + 1] = query[:query_len], query_len
+                    X2[i * 2 + 1, :neg_doc_len], X2_len[i * 2 +
+                                                        1] = neg_doc[:neg_doc_len], neg_doc_len
```

Comment:
This formatting looks like some ASCII-art poem :)

Reply:
Haha
I have a Sublime Text plugin for making it auto PEP8 compliant.
I will format it manually.